src/env.h

```diff
@@ -520,10 +520,14 @@ struct ContextInfo {
   bool is_default = false;
 };
 
-struct CompileFnEntry {
+struct CompiledFnEntry {
   Environment* env;
   uint32_t id;
-  CompileFnEntry(Environment* env, uint32_t id);
+  v8::Global<v8::Function> function;
```

Comment:
@hashseed will `function` prevent `script` from being collected? `script` is weak and `function` is not. I didn't notice any leaks while testing but it feels odd.

Reply:
`v8::Function` is an alias for `v8::internal::JSFunction`. I'd expect it to reference the script object through `internal_function->shared()->script()`, i.e., `v8::internal::SharedFunctionInfo::shared()`.

Reply:
@bnoordhuis that's why I'm concerned. my first thought was this might be an unresolvable cycle if V8 can't tell they're part of the same CompiledFnEntry, but I couldn't detect any leaks.

Reply:
@devsnek I think it would be weird if keeping the script alive from a function works for scripts compiled with `vm.runInThisContext()` but not here (although I still don’t understand the V8 internals here well – `internal_function->shared()->script()` does not seem to correspond to the public `v8::Script`).

Ultimately, the function here is only used as a map key for `callbackMap`, so I think it would be okay to create a dummy object from `vm.compileFunction()` and use that as the cache key instead of the function?

Reply:
when i replace function with a dummy object this goes back to not working :( i think we may need a bit more help from the v8 side.

Reply:
I think that would mean that this only works because it does, in fact, inhibit GC altogether currently?

Reply:
> `internal_function->shared()->script()` does not seem to correspond to the public `v8::Script`

Oh, I guess that's right. V8 treats scripts as functions internally and a `v8::Script` is really a reinterpret_casted `v8::internal::JSFunction`. I suppose that means the current logic is okay.